concert/ext/tangoservers/porqa.py

```diff
@@ -0,0 +1,141 @@
+        z, y, x = np.indices(fft1.shape)
+        center: ArrayLike = np.array(fft2.shape) // 2
+        rz, ry, rx = z - center[0], y - center[1], x - center[2]
+        # Directional distance
+        if direction == 'xy':
+            r: float = np.sqrt(rx**2 + ry**2)
+        elif direction == 'yz':
+            r: float = np.sqrt(ry**2 + rz**2)
+        elif direction == 'zx':
+            r: float = np.sqrt(rz**2 + rx**2)
+        else:
+            raise ValueError("Invalid direction")
+        r = r.astype(int)
+        fsc_num: ArrayLike = np.zeros(self._freq_bins)
+        fsc_den1: ArrayLike = np.zeros(self._freq_bins)
+        fsc_den2: ArrayLike = np.zeros(self._freq_bins)
+        for i in tqdm(range(self._freq_bins), total=self._freq_bins):
+            mask = (r == i)
+            fft1_shell = fft1[mask]
+            fft2_shell = fft2[mask]
+            if len(fft1_shell) > 0:
+                fsc_num[i] = np.sum(fft1_shell * np.conj(fft2_shell)).real
+                fsc_den1[i] = np.sum(np.abs(fft1_shell)**2)
+                fsc_den2[i] = np.sum(np.abs(fft2_shell)**2)
+        fsc: ArrayLike = fsc_num / (np.sqrt(fsc_den1 * fsc_den2) + 1e-8)
+        self.info_stream("%s: SFSC in %s direction : %s", self.__class__.__name__, direction, fsc)
+        return fsc
+
+    async def _exec_fsc(self, producer: AsyncIterator[ArrayLike]) -> None:
+        """Computes Self Fourier Shell Correlation (FSC) for reconstructed volume"""
+        slices: List[ArrayLike] = [slice async for slice in producer]
+        volume: ArrayLike = np.stack(slices, axis=0)
+        if volume.ndim != 3:
+            self.error_stream(
+                "%s: expected 3D volume, got %d", self.__class__.__name__, volume.ndim)
+            return
+        self.info_stream("%s: computing directional SFSC with %d frequency bins",
+                         self.__class__.__name__, self._freq_bins)
+        even_volume, odd_volume = volume[::2, :, :], volume[1::2, :, :]
+        min_z = min(even_volume.shape[0], odd_volume.shape[0])
+        even_volume, odd_volume = even_volume[:min_z], odd_volume[:min_z]
+        self.info_stream("%s: dtsributed volumes into even and odd slices having shape: %s",
+                         self.__class__.__name__, even_volume.shape)
+        xp._default_memory_pool.free_all_blocks()
+        # Compute Fourier transforms
+        even_fft: ArrayLike = xp.asnumpy(
+            xft.fftshift(xft.fftn(xp.asarray(even_volume.astype(np.float32)))))
+        odd_fft: ArrayLike = xp.asnumpy(
+            xft.fftshift(xft.fftn(xp.asarray(odd_volume.astype(np.float32)))))
+        xp._default_memory_pool.free_all_blocks()
+        fsc_xy: ArrayLike = self._compute_directional_fsc(even_fft, odd_fft, direction='xy')
+        fsc_yz: ArrayLike = self._compute_directional_fsc(even_fft, odd_fft, direction='yz')
+        fsc_zx: ArrayLike = self._compute_directional_fsc(even_fft, odd_fft, direction='zx')
```

Comment:
It would be good to also compute the 3D shell correlation, not only the cross-sections.